Implementation of SOSDacApi GetMethodDescName for cDAC

davidwrighton opened the pull request:
Add a number of new `MethodDesc` contract definitions

| Contract algorithm on RuntimeTypeSystem |
| --- |
| `IsGenericMethodDefinition`|
|`GetGenericMethodInstantiation`|
|`GetMethodToken`|
|`IsArrayMethod`|
|`IsDynamicMethod`|
|`IsStoredSigMethodDesc`|
|`IsNoMetadataMethod`|
|`IsILStub`|

Update cDAC compat asserts in cDAC to always be enabled by using a tls variable in `mscordaccore`

Implement `GetMethodDescName` on `ISOSDacInterface` in the `cdacreader`

Stub out an implementation of `GetPath` in the `Loader` contract used in a fallback after a fallback. This will need further work, but is included to make sure the code path isn't lost.

Fix the `EcmaMetadataReader` to be able to find blobs in the metadata

Add ability to read target data from a buffer held on the cdac side using the `Target` class. This was needed to handle signature containing a `CorElementType.Internal`.

And finally actually implement the name generation algorithm via a line for line port from the CoreCLR codebase.

Contributes to https://github.com/dotnet/runtime/issues/99302

lambdageek committed:
start GetMethodDescDataImpl

lambdageek committed:
WIP: managed GetMethodDescData skeleton

davidwrighton committed:
Progress towards GetMethodTableName

lambdageek committed:
wip: MethodDesc

lambdageek committed:
add MethodDescChunk

davidwrighton committed:
Move metadata reader into helpers
Add implementations for changes to RuntimeTypeSystem contract
Add SOSDacInterface call into cDac for GetMethodTableName

davidwrighton committed:
Simplify storage of saved metadata copy for reflection emit scenarios
- Move pointer to Module class
- Replace use of SBuffer abstraction with a simple counted byte memory block
Add metadata details to Loader contract
Add Metadata helper api for use by contracts within the cDAC (and possibly clients of cDAC too)

lambdageek committed:
WIP: validating a MethodDesc

davidwrighton committed:
Fix bugs found when trying to actually use this logic
1. It works now!

davidwrighton committed:
Update contract documentation

davidwrighton committed:
Merge branch 'main' into cdac-methodtable-name

lambdageek committed:
checkpoint: MethodDesc validation

lambdageek committed:
update contract

davidwrighton committed:
Feedback

lambdageek committed:
fix RuntimeTypeSystem unit tests

mock the additional data and globals

davidwrighton committed:
Add support for the EEName DacStream

lambdageek committed:
update contract

lambdageek committed:
fix GetMethodDescChunkPointerMayThrow

davidwrighton committed:
- Refactor to expose TypeHandle exclusively from the contracts
- Remove TypeHandleArray and MethodTableArray in favor of contract specific logic

davidwrighton committed:
Address concern around data structure for DynamicMetadata

davidwrighton committed:
Merge branch 'cdac-methodtable-name' of github.com:davidwrighton/runtime into cdac-methodtable-name

davidwrighton committed:
Fix build break

davidwrighton committed:
Fix musl build issue

lambdageek committed:
Cherrypick RuntimeTypeSystem changes from #104759

Co-Authored-By: David Wrighton <davidwr@microsoft.com>

lambdageek committed:
update tests

lambdageek committed:
update tests

lambdageek committed:
fixup rebase

davidwrighton committed:
Remove unnecessary usings

davidwrighton committed:
- Transform magic numbers from DacStreams implementation into constants
- Add test suite for the DacStreams contract

davidwrighton committed:
Feedback

davidwrighton committed:
Merge branch 'cdac-methodtable-name' into cdac-methoddescname

davidwrighton committed:
Add non-contract portion of GetMethodDescName

davidwrighton committed:
Remove default parameter from AppendMethodInternal

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into cdac-methoddescname

davidwrighton committed:
First stab at the actual contract implementations for the MethodDesc reading needs

davidwrighton committed:
Commit current state

davidwrighton committed:
It Works!

davidwrighton committed:
Document the new contracts needed for generating MethodDesc names

davidwrighton committed:
Handle Path fallback scenario. Leave actual implementation/documentation of contract Loader.GetPath for a future date

davidwrighton committed:
Remove extra lines found in Loader_1.cs

lambdageek commented on docs/design/datacontracts/RuntimeTypeSystem.md:
````diff
@@ -572,12 +610,211 @@ will typically have multiple chunks.  There are subkinds of MethodDescs at runti
+```csharp
+    internal enum MethodDescClassification
+    {
+        mcIL = 0, // IL
+        mcFCall = 1, // FCall (also includes tlbimped ctor, Delegate ctor)
+        mcNDirect = 2, // N/Direct
+        mcEEImpl = 3, // special method; implementation provided by EE (like Delegate Invoke)
+        mcArray = 4, // Array ECall
+        mcInstantiated = 5, // Instantiated generic methods, including descriptors
+                            // for both shared and unshared code (see InstantiatedMethodDesc)
+        mcComInterop = 6,
+        mcDynamic = 7, // for method desc with no metadata behind
````
nit: I've been leaving off prefixes on C# enums when bringing them over from C++

lambdageek commented on src/native/managed/cdacreader/src/Contracts/RuntimeTypeSystem_1.cs:
```diff
@@ -71,26 +73,77 @@ internal enum TypeHandleBits
+        {
+            get
+            {
+                int tokenRemainderBitCount = _target.ReadGlobal<byte>(Constants.Globals.MethodDescTokenRemainderBitCount);
```
Can we either store the token remainder bit count in the MethodDesc, or just pre-compose the token when creating the MethodDesc instance?  I've been avoiding storing `Target` and contracts in data structures, so they're just dumb value holders, not fancy algorithms.

lambdageek commented on src/native/managed/cdacreader/src/Contracts/RuntimeTypeSystem_1.cs:
```diff
@@ -487,10 +540,194 @@ public MethodDescHandle GetMethodDescHandle(TargetPointer methodDescPointer)
+
+        if (methodDesc.Classification != MethodDescClassification.mcInstantiated)
+            return false;
+
+        Data.InstantiatedMethodDesc instantiatedMethodDesc = _target.ProcessedData.GetOrAdd<Data.InstantiatedMethodDesc>(methodDescHandle.Address);
+        return ((int)instantiatedMethodDesc.Flags2 & (int)InstantiatedMethodDescFlags2.KindMask) == (int)InstantiatedMethodDescFlags2.GenericMethodDefinition;
+    }
```
I think we can be a bit more high level here:

I really don't like this pattern for the MethodDesc subclasses.

I've been doing this in my fork:
```
    // member of RuntimeTypeSystem_1
    internal InstantiatedMethodDesc AsInstantiatedMethodDesc(MethodDesc md)
    {
        Debug.Assert(md.Classification == MethodClassification.Instantiated);
        Data.InstantiatedMethodDesc imd = _target.ProcessedData.GetOrAdd<Data.InstantiatedMethodDesc>(md.Address);
        return new InstantiatedMethodDesc(md.Address, imd);
    }
```

where `InstantiatedMethodDesc` is a wrapper around `Data.InstantiatedMethodDesc`:

```csharp
  // also in RuntimeTypeSystem_1
      internal struct InstantiatedMethodDesc
    {
        private readonly TargetPointer _address;
        private readonly Data.InstantiatedMethodDesc _desc;
        internal InstantiatedMethodDesc(TargetPointer address, Data.InstantiatedMethodDesc instantiatedMethodDesc)
        {
            _address = address;
            _desc = instantiatedMethodDesc;
        }

        private bool HasFlags(InstantiatedMethodDescFlags mask, InstantiatedMethodDescFlags flags) => (_desc.Flags2 & (ushort)mask) == (ushort)flags;
        internal bool IsWrapperStubWithInstantiations => HasFlags(InstantiatedMethodDescFlags.KindMask, InstantiatedMethodDescFlags.WrapperStubWithInstantiations);
        internal bool IsGenericMethodDefinition => HasFlags(InstantiatedMethodDescFlags.KindMask, InstantiatedMethodDescFlags.GenericMethodDefinition);

        internal bool HasPerInstInfo => _desc.PerInstInfo != TargetPointer.Null;

        internal bool HasMethodInstantiation => IsGenericMethodDefinition || HasPerInstInfo;

    }
```

lambdageek reviewed (commented):
First set of comments.  Didn't look at Ecma or the formatter yet

lambdageek commented on src/native/managed/cdacreader/src/Target.cs:
```diff
@@ -5,6 +5,7 @@
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
+using System.Net;
```
nit: unused (presmably) using

lambdageek reviewed (approved):
LGTM. modulo nits.  And it would be really really nice if `RuntimeTypeSystem_1.MethodDesc` didn't hold on to `Target`

davidwrighton committed:
Address code review comments

1. Rename mcNDirect to mcPInvoke
2. Rename MethodDescClassification to MethodClassification
3. Remove mc prefix from all MethodClassification enum values
4. Move handling of various forms of MethodDescs to using an AsBlah method, and handle the flags in those methods, to make the actual contract methods somewhat simpler.
5. Compute the token eagerly on MethodDesc creation to avoid having to hold a pointer to the Target

jkotas commented on docs/design/datacontracts/RuntimeTypeSystem.md:
```diff
@@ -92,12 +92,49 @@ struct MethodDescHandle
+
+    // Return true for a MethodDesc that describes a method represented by the System.Reflection.Emit.DynamicMethod class
+    // A LCG method is also a StoredSigMethodDesc
+    public virtual bool IsLCGMethod(MethodDescHandle methodDesc);
```
`DynamicMethod` is the public name of the LCG feature.

ReflectionEmit is more general. Some methods emitted using Reflection.Emit are dynamic methods, some are not.

davidwrighton replied:
... whereas I think of Reflection.Emit as a thing that only applies to things that generate the full metadata and types which is also not completely accurate...   The actual term in the BCL is that an LCGMethod maps to a method represented by the `System.Reflection.Emit.DynamicMethod` class, but calling it a DynamicMethod... might be extra special confusing, since the runtime has multiple meanings of what a `DynamicMethodDesc` represents, only 1 of which is a `System.Reflection.Emit.DynamicMethod`. (And of course LCG stands for Lightweight CodeGen, which is an internal codename that probably doesn't still need to be kept around.) I'd like to hear what @lambdageek, and @AaronRobinsonMSFT  think on this, as I know I'm not objective here as I've worked on this stuff for FAR too long to see things reasonably. Possibly the right approach would be to call it `DynamicMethod` and sometime in .NET 10, rename the `DynamicMethodDesc` to something like `RuntimeILMethodDesc`.

AaronRobinsonMSFT replied:
I like the rename that elinor mentioned above at https://github.com/dotnet/runtime/pull/106169#discussion_r1712449039.

> rename the `DynamicMethodDesc` to something like `RuntimeILMethodDesc`.

I also agree with your suggestion above. I could also see `GeneratedMethodDesc` or `ILEmitMethodDesc`.

lambdageek replied:
I like `IsLCGMethod` -> `IsDynamicMethod` and `DynamicMethodDesc` ->`ILEmitMethodDesc`.

Not sure about `IsDynamicMethod` -> `IsDynamicallyGeneratedMethod`. Would just `IsGeneratedILMethod` be too broad?

jkotas replied:
> DynamicMethodDesc ->ILEmitMethodDesc.

How about `DynamicMethodDesc` ->`NoMetadataMethodDesc` and `IsDynamicMethod` -> `IsNoMetadata` (existing property)?

Part of the codebase uses the NoMetadata name already: 
https://github.com/dotnet/runtime/blob/fa33eddf3533dac554e959da92f54104c0568c1e/src/coreclr/vm/method.hpp#L626-L630 . Notice that `MethodDesc::IsDynamicMethod` and `MethodDesc::IsNoMetadata` have the same implementation.

The key property of the `DynamicMethodDesc` is that it has no ECMA-335 metadata, it has no metadata token, etc.

Dynamic IL generation is not the key property of `DynamicMethodDesc`. The regular ECMA-335 MethodDescs can have dynamically generated (IL) code too. For example, `UnsafeAccessors` have dynamically generated IL but they are regular MethodDescs.

davidwrighton replied:
I like @jkotas's idea here. Then the IsDynamicMethod api at the DAC level makes total sense, the existing `DynamicMethodDesc` translates to `NoMetadataMethodDesc` and gets a better name that more matches its utility, and we get rid of having both `MethodDesc::IsNoMetadata` and `MethodDesc::IsDynamicMethod` which are today exactly the same thing. While I'm at it, I'll rename `mcDynamic` to `mcNoMetadata`

davidwrighton replied:
I'm trending toward the idea of putting this rename in a separate PR. Opinions?

davidwrighton replied:
Offline discussion is to put together a separate PR for that. I'll just add a commit to update the cdac contract side naming to IsDynamicMethod, and the renaming of the underlying structures to be less confusing will be a separate PR that will probably miss .NET 9.

davidwrighton commented on docs/design/datacontracts/RuntimeTypeSystem.md:
```diff
@@ -92,12 +92,49 @@ struct MethodDescHandle
+    // Return true if a MethodDesc represents a dynamically generated method, either an IL Stub dynamically
+    // generated by the runtime, or a MethodDesc that describes a method represented by the System.Reflection.Emit.DynamicMethod class
+    // A dynamic method is also a StoredSigMethodDesc
+    public virtual bool IsDynamicMethod(MethodDescHandle methodDesc, out ReadOnlySpan<byte> methodName);
```
Not a bad idea. In general, I don't want to expose arrays to contract users, as its far to easy to mutate them, but strings are nice in that they are our only truly immutable type in the BCL, and work nicely for this sort of thing.

davidwrighton committed:
Adjust cdac naming to new model

davidwrighton committed:
Merge branch 'main' into cdac-methoddescname
